File: app/src/main/java/net/ginteam/carmen/presenter/city/CitiesPresenter.java
package net.ginteam.carmen.presenter.city;

import net.ginteam.carmen.contract.city.CityContract;
import net.ginteam.carmen.model.city.CityModel;
import net.ginteam.carmen.provider.ModelCallback;
import net.ginteam.carmen.provider.city.CitiesProvider;

import java.util.List;

/**
 * Created by Eugene on 12/27/16.
 */

public class CitiesPresenter implements CityContract.Presenter {

    private CityContract.View mView;

    @Override
    public void selectCity(CityModel city) {

    }

    @Override
    public void fetchData() {
        mView.showLoading(true);

        CitiesProvider
                .getInstance()
                .fetch(new ModelCallback<List<CityModel>>() {
                    @Override
                    public void onSuccess(List<CityModel> resultModel) {
                        mView.showLoading(false);
                        mView.showCities(resultModel);
                    }

                    @Override
                    public void onFailure(String message) {
                        mView.showError(message);
                    }
                });
    }

    @Override
    public void attachView(CityContract.View view) {
        mView = view;
    }

    @Override
    public void detachView() {
        mView = null;
    }

}

File: app/src/main/kotlin/net/ginteam/carmen/kotlin/presenter/city/CitiesPresenter.kt
package net.ginteam.carmen.kotlin.presenter.city

import net.ginteam.carmen.kotlin.api.response.ModelSubscriber
import net.ginteam.carmen.kotlin.contract.CitiesContract
import net.ginteam.carmen.kotlin.manager.PreferencesManager
import net.ginteam.carmen.kotlin.manager.SharedPreferencesManager
import net.ginteam.carmen.kotlin.model.CityModel
import net.ginteam.carmen.kotlin.presenter.BasePresenter
import net.ginteam.carmen.kotlin.provider.CitiesDataSourceProvider
import net.ginteam.carmen.kotlin.provider.OnlineCitiesDataSourceProvider

/**
 * Created by eugene_shcherbinock on 2/16/17.
 */

class CitiesPresenter : BasePresenter <CitiesContract.View>(), CitiesContract.Presenter {

    private val mCitiesProvider: CitiesDataSourceProvider = OnlineCitiesDataSourceProvider()

    override fun fetchCities() {
        mView?.showLoading(true)

        mCitiesProvider
                .fetchCities()
                .subscribe(object : ModelSubscriber <List <CityModel>>() {
                    override fun success(model: List<CityModel>) {
                        mView?.showLoading(false)
                        mView?.showCities(model)
                    }

                    override fun error(message: String) {
                        mView?.showError(message)
                    }
                })
    }

    override fun saveSelectedCity(city: CityModel) {
        val preferences: PreferencesManager = SharedPreferencesManager
        preferences.userCityModel = city
    }
}